assets/files/usr/local/bin/haproxy.sh

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env bash
+set -e
+
+function etcd_members {
+    local domain
+
+    domain="$1"
+    dig +noall +answer -t SRV "_etcd-server-ssl._tcp.$domain" | awk '{print $NF}'
+}
+
+function first_a_addr {
+    local name
+
+    name="$1"
+    dig +noall +answer "$name"  | awk '$4 == "A" {print $NF; exit}'
+}
+
+function generate_haproxy_cfg {
+    local ip
+    local template_path
+    local cfg_path
+    local domain
+    local api_port
+
+    domain="$1"
+    template_path="${2}/haproxy.cfg.template"
+    cfg_path="${2}/haproxy.cfg"
+    api_port="$3"
+
+    sudo tee "$cfg_path" < "$template_path"
+    for item in $(etcd_members "$domain"); do
+        if ip=$(first_a_addr "$item") && [[ -n "$ip" ]]; then
+            sudo echo "   server $item ${ip}:$api_port weight 1 check inter 3s fall 3 rise 3" | sudo tee -a "$cfg_path"
+        fi
+    done
+}
+
+function has_master_api_lb_topology_changed {
+    local ip
+    local haproxy_cfg_ip
+    local domain
+    local cfg
+
+    domain="$1"
+    cfg="$2"
+    for item in $(etcd_members "$domain"); do
+        ip=$(first_a_addr "$item")
+        haproxy_cfg_ip=$(awk -v "server=$item" '$2 == server {print substr($3, 0, index($3, ":") - 1); exit}' "$cfg")
+        if [[ $? -eq 0 && -n "$ip" ]]; then
```

Comment:
The '$? -eq 0' refers to the operation in line #48, right?
In what case, we'll get an error (!=0) for that line?